---
author: "Pol Sole-Navais"
date: "`r paste('Analysis date:', format(Sys.time(), '%d %B, %Y'))`"
output: tint::tintPdf
link-citations: yes
latex_engine: pandoc
---

```{r echo=F, include= F}
library(tint)

knitr::opts_chunk$set(tidy = FALSE, cache.extra = packageVersion('tint'))
options(htmltools.dir.version = FALSE)
pheno= snakemake@wildcards[['pheno']]
pdf.options(useDingbats = TRUE)
#knitr::opts_chunk$set(dpi=300)
knitr::opts_chunk$set(dev = 'png', warning= FALSE, message= FALSE, dpi= 600)


pheno= ifelse(pheno=='allPTD', 'Preterm Delivery', ifelse(pheno== 'postTerm', 'Post Term', ifelse(pheno=='GAraw', 'Gestational duration', 'Normalized Gestational Duration')))

```

---
title: "`r paste('Forest plots for ', pheno)`"
---

```{r include=FALSE, echo=F}
library("ggplot2")
library("dplyr")
library("knitr")
library("tidyr")
library(cowplot)
library("kableExtra")
library(ggrepel)
library("data.table")
options(warn=-1)

colorBlindBlack8  <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
```


```{r echo= F}
d= fread(snakemake@input[[1]])

z= fread(snakemake@input[[3]])




df= fread(snakemake@input[[2]], select= (c('MarkerName', 'Effect', 'StdErr', 'HetISq', 'HetPVal', 'TOTALSAMPLESIZE', 'P-value', 'Allele1', 'Allele2')))
names(df)= c('SNP', 'BETA', 'SE', 'HetISq', 'HetPval', 'N', 'pvalue', 'A1', 'A2')
df= filter(df, SNP %in% d$SNP)

df= separate(df, SNP, into= c('CHR', 'POS', 'Ax1', 'Ax2', 'ID'), sep= ':', remove= F)
df$BETA= ifelse(df$A2 > df$A1, -1 * df$BETA, df$BETA)
df$CHR= ifelse(df$CHR== 'X','23', df$CHR)
df$CHR= as.integer(df$CHR)
df$POS= as.integer(df$POS)
df= select(df, -c(A1, A2, ID, Ax1, Ax2))

df$cohort= 'Meta-analysis'
d= bind_rows(d, df)

z$CHR= ifelse(z$CHR== 'X','23', z$CHR)
z$CHR= as.integer(z$CHR)

d= inner_join(d, z, by= 'CHR') %>% filter(POS> pos1, POS< pos2)



d$locus= paste0('Chr ', d$CHR,': ', d$nearestGene)
d$cohort= paste0(d$cohort, ' (n= ', d$N, ')')

```

```{r, echo=F, results='asis', fig.align = "center", fig.width= 11, fig.fullwidth=TRUE, fig.height= 7, dpi= 600}

for (i in unique(d$locus)) {
temp_df= d[d$locus== i, ]

cat('\n')

cat("\n# Forest plot for locus ", i, "\n")

cat("\n")

cat('\n')

cat('Lead variant: \n', temp_df[!is.na(temp_df$HetISq), ]$SNP)

cat('\n')

cat(paste0('\n Meta-analysis: Beta= ', round(temp_df[!is.na(temp_df$HetISq), ]$BETA, 3), ' (95% CI= ', round(temp_df[!is.na(temp_df$HetISq), ]$BETA - 1.96 * temp_df[!is.na(temp_df$HetISq), ]$SE, 3), ', ', round(temp_df[!is.na(temp_df$HetISq), ]$BETA + 1.96 * temp_df[!is.na(temp_df$HetISq), ]$SE, 3), '); pvalue= ', temp_df[!is.na(temp_df$HetISq), ]$pvalue))

cat('\n')

cat('\n')
temp_df= temp_df[order(temp_df$N, decreasing= T), ]

(ggplot(temp_df, aes(x=factor(cohort, level = factor(cohort)), y=BETA, ymin= BETA - 1.96 * SE, ymax= BETA + 1.96 * SE, colour= !is.na(HetISq), shape= !is.na(HetISq)), alpha= 0.5) +
 geom_pointrange(size= 1, alpha= 0.7) +
scale_shape_manual(values= c(15, 18), guide= F) +
 geom_hline(yintercept = 0, linetype=2) +
scale_y_continuous(sec.axis = dup_axis()) +
 coord_flip() +
scale_colour_manual(values= c(colorBlindBlack8[3], colorBlindBlack8[4]), guide= F) +
theme_cowplot() +
 xlab('') +
    ylab('Beta [95% CI]') +
geom_vline(xintercept= 0, linetype= "dotted", colour= 'grey') ) %>% print()

cat('\n')

cat('Test for heterogeneity: I^2^= ', temp_df[!is.na(temp_df$HetISq), ]$HetISq, '%; Het pvalue= ', temp_df[!is.na(temp_df$HetISq), ]$HetPval, '\n')

cat('\\pagebreak')

}

```


