---
author: "Pol Sole-Navais"
date: "`r paste('Report date:', format(Sys.time(), '%d %B, %Y'))`"
output: tint::tintPdf
link-citations: yes
latex_engine: xelatex
---

```{r echo=F, include= F}
library(tint)

knitr::opts_chunk$set(tidy = FALSE, cache.extra = packageVersion('tint'))
options(htmltools.dir.version = FALSE)
pheno= snakemake@wildcards[['pheno']]
pdf.options(useDingbats = TRUE)
#knitr::opts_chunk$set(dpi=300)
knitr::opts_chunk$set(dev = 'png', warning= FALSE, message= FALSE, dpi= 600)


pheno= ifelse(pheno=='allPTD', 'Preterm Delivery', ifelse(pheno== 'postTerm', 'Post Term', ifelse(pheno=='GAraw', 'Gestational duration', 'Normalized Gestational Duration')))

```

---
title: "`r paste('EGG GWAS meta-analysis of', pheno, 'using maternal genome.')`"
#subtitle: "`r paste('Preliminary results on', pheno)`"
---



```{r include=FALSE, echo=F}
library("ggplot2")
library("dplyr")
library("knitr")
library("tidyr")
library(cowplot)
library("kableExtra")
library(ggrepel)
library("data.table")
options(warn=-1)

```

```{r echo=F, fig.margin = TRUE, fig.cap = 'Density plot of minor allele frequency.'}

d= fread(snakemake@input[[1]], h= T)
d$MAF= ifelse(d$EAF>0.5,  1 - d$EAF, d$EAF)

colorBlindBlack8= c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

ggplot(d, aes(MAF)) +
  geom_density(fill= colorBlindBlack8[4]) +
theme_cowplot(font_size= 12) +
xlab('MAF')

#d= select(d, -MAF)

```

```{r echo=F, fig.margin = TRUE, fig.cap = 'Density plot of sample size per genetic variant.'}


ggplot(d, aes(TOTALSAMPLESIZE)) +
  geom_density(fill= colorBlindBlack8[3]) +
theme_cowplot(font_size= 12) +
xlab('Sample size')


```

```{r echo=F, fig.margin = TRUE, fig.cap = 'Density plot of beta estimates.'}


ggplot(d, aes(BETA)) +
  geom_density(fill= colorBlindBlack8[2]) +
theme_cowplot(font_size= 12) +
xlab('Beta')


```



# Summary  

We conducted fixed effect inverse-variance weighted meta-analysis of `r pheno` on `r max(d$TOTALSAMPLESIZE)` women using METAL. This report contains the results obtained after QC. A total of `r nrow(d)` genetic variants are described in this report.  

For binary traits, we only included cohorts with an effective sample size> 100.  

```{marginfigure}
**Effective sample size** for binary phenotypes was calculated as:  
$$\frac{2}{(\frac{1}{Ncases} + \frac{1}{Ncontrols})}$$  

```

Per-study genetic variants were included if:  
- MAF> 0.5%  
- MAC> 6  
- Allele matching with HRC or 1KG (restricted to white Europeans)  
- Imputation score> 0.4  
- 0 < P-value < 1  
- Expected pvalue (calculated from z-score) and reported pvalue do not differ more than 10%  
- Effect allele frequency differs less than 0.2 compared to HRC or 1KG (restricted to white Europeans)  





```{r, echo=F}
ge= data.frame(CHR= c(5, 3, 1, 23, 1), pos_ge= c(157895049, 127881613, 22470407, 115164770, 22470407))

#kable(summary(select(d, BETA, SE, pvalue, EAF, TOTALSAMPLESIZE)), digits = c(3, 3, 5, 4, 0), col.names = c('Beta', 'Standard error', 'P-value', 'Effect allele frequency', 'Sample size'), caption= 'Summary statistics after QC.')
```

\newpage  


# Manhattan plot  


We identified independent loci as genetic variants within a radius of 250kb from the lead SNP (p-value< 5x10^-8^). Previous discoveries for either gestational duration or preterm delivery were obtained from the previous largest [GWAS](https://www-nejm-org.ezproxy.ub.gu.se/doi/full/10.1056/nejmoa1612665) on the outcome.  

```{marginfigure}
The same number of loci is obtained when using a larger radius (1.5Mb).
```

```{marginfigure}
We note that we used a naive approach to identify independent loci. This should be interpreted cautiously. We mapped top genetic variants to the body (TSS or TES) nearest protein coding gene.  

```


```{r, echo=F, fig.align = "center", fig.width= 10, fig.fullwidth=TRUE, fig.height= 4, dpi= 600, fig.cap= 'Manhattan plot'}

df= arrange(d, pvalue)

#df= group_by(df, CHR, POS) %>% filter(row_number() == 1) %>% ungroup()
df= df[!duplicated(df[, c('CHR', 'POS')]), ]

dg= df %>% arrange(CHR, POS) %>% filter(pvalue< 5*10**-8) %>% group_by(CHR) %>%
		mutate(d=POS-lag(POS, default=-Inf), clumpid=cumsum(d>250000)) %>%
		group_by(CHR, clumpid) %>%
		filter(rank(pvalue, ties.method = "random")==1) %>%
		mutate(GENE= ifelse(SYMBOL=='', RSID, SYMBOL)) %>%
		ungroup()

dg= filter(dg, pvalue< 5*10**-8)
dg= group_by(dg, CHR, POS) %>% filter(row_number()== 1)
#dg$GENE= ifelse(grepl('rs|-', dg$GENE), dg$nearestGene, ifelse(dg$GENE=='', dg$nearestGene, dg$GENE))
dg$GENE= dg$nearestGene

  don <- df %>%
    group_by(CHR)      %>%
    summarise(chr_len= max(POS)) %>%
    mutate(tot= cumsum(as.numeric(chr_len))-chr_len) %>% # Calculate cumulative position of each chromosome
    select(-chr_len) %>%
    left_join(df, ., by= 'CHR') %>%
    arrange(CHR, POS) %>% # Add a cumulative position of each SNP
    mutate( BPcum=POS+tot) %>%
	 ungroup()

  axisdf = don %>% group_by(CHR) %>% summarize(center=( max(BPcum) + min(BPcum) ) / 2 )
  names(axisdf)= c('CHR', 'center')
HC= -log10(5*10**-8)
dg= dg %>% ungroup() %>% select(ID, GENE, CHR, POS, MAF, BETA)
don= left_join(don, select(dg, ID, GENE), by= 'ID')
names(dg)= c('ID', 'GENE', 'CHR', 'POS_new', 'MAF', 'BETA')

lims= 250000

don$disc= ifelse(don$pvalue> 5*10**-8, 0, NA)

for (i in rownames(dg)) {
don= mutate(don, disc= ifelse(CHR== as.integer(dg[i, 'CHR']) & POS>= as.integer(dg[i, 'POS_new']) - lims & POS<= as.integer(dg[i, 'POS_new']) + lims, 2, disc))
}

for (i in rownames(ge)) {
don= mutate(don, disc= ifelse(CHR== as.integer(ge[i, 'CHR']) & POS>= as.integer(ge[i, 'pos_ge']) - lims & POS<= as.integer(ge[i, 'pos_ge']) + lims, 1, disc))
}

don= don[order(don$disc, decreasing= F, na.last= T), ]
don$disc= factor(don$disc, levels=c(0, 1, 2), labels=c('Not significant', 'Previous discovery', 'New discovery'))

cols <- c('Not significant'= 'grey', 'Previous discovery'= colorBlindBlack8[4], 'New discovery'= colorBlindBlack8[2])

don$GENE= ifelse(!is.na(don$GENE), don$nearestGene, don$GENE)
ggplot(don) +
    geom_point(data= don, aes(x=BPcum, y= -log10(pvalue), colour= disc), size=0.3) +   # Show all points
theme_cowplot(font_size= 12) + #theme_minimal_hgrid(12, rel_small = -1) + 
#scale_alpha_manual(values= rep(c(1/10, 1/2), 23)) +
scale_colour_manual(values= cols) +
    scale_x_continuous(label = axisdf$CHR, breaks= axisdf$center, expand=c(0,0) ) + # custom X axis
scale_y_continuous(expand= c(0,0)) +
         xlab('Chromosome') +
    ylab('-log10(pvalue)') +
labs(colour= '') +
geom_hline(yintercept= 0, size= 0.5, colour= 'black') +
geom_hline(yintercept= HC, size= 0.5, linetype= 2, colour= '#878787') +
geom_text_repel(data= don, aes(x= BPcum, y= -log10(pvalue), label= GENE), size= 3, hjust= 1, force= 1, vjust= 1, colour= 'black') +
theme(legend.position= 'bottom') +
guides(colour = guide_legend(override.aes = list(size=3)))


```

```{r, echo=F, fig.align = "center", fig.width= 10, fig.fullwidth=TRUE, fig.height= 4, dpi= 600, fig.cap= 'Independent signals within each cluster'}

indep= fread(snakemake@input[[6]])

indep$nd2P= sapply(strsplit(indep$SP2, ','), length)
indep= inner_join(indep, dg, by= 'CHR')
indep= filter(indep, BP>= POS_new - 1.5*10**6, BP<= POS_new + 1.5*10**6)
indep_df= group_by(indep, GENE) %>% summarize(total= sum(TOTAL), nsig= sum(NSIG), GWS= n(), sug_ev= sum(nd2P), mP= min(P))
indep_df= indep_df[order(indep_df$mP, decreasing=T), ]
indep_df$GENE= factor(indep_df$GENE, levels= indep_df$GENE)
indep_df= filter(indep_df, !grepl('HLA', GENE))

p1= ggplot(data=indep_df, aes(x= GENE, y= GWS)) + 
geom_col(fill=colorBlindBlack8[2], alpha= 0.6) + 
theme_cowplot() +
ylab('# Independent GW significant') +
xlab('Locus') +
coord_flip()

p2= ggplot(data=indep_df, aes(x= GENE, y= total)) + 
geom_col(fill=colorBlindBlack8[4], alpha= 0.6) + 
theme_cowplot() +
ylab('Total # of \n genetic variants in locus') +
xlab('Locus') +
theme(axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank()) +
coord_flip() 

p3= ggplot(data=indep_df, aes(x= GENE, y= sug_ev / total)) + 
geom_col(fill=colorBlindBlack8[8], alpha= 0.6) + 
theme_cowplot() +
ylab('Proportion of variants with P<1e-5') +
xlab('Locus') +
ylim(0, 1) +
theme(axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank()) +
coord_flip() 

plot_grid(p1, p2, p3, align = "h", nrow= 1)

```



We have purposedly left out the HLA region from this analysis.  
Clumps were obtained using the following plink command:  
plink --bfile {params[0]} --clump {input[0]} --exclude {input[1]} --clump-r2 0.05 --clump-kb 1000 --clump-p1 5e-8 --clump-p2 1e-5 --out {params[1]}

That is, we only kept index variants with a p-value< 5e-8, and suggestive hits were tagged using a p-value cut-off < 1e-5. Variants which are less than 1000 Kb away from the index variant and have R2 larger than 0.05 with it are assigned to that index variant's clump.  



# QQplot by minor allele frequency  

```{r, echo=F, fig.align = "center", fig.width= 5, dpi= 600, fig.cap= 'QQ-plot by MAF tertiles'}

d= mutate(d, maf_tertiles = ntile(MAF, 3))
m1= round(max(d[d$maf_tertiles== 1, 'MAF']), 3)
m2= round(max(d[d$maf_tertiles== 2, 'MAF']), 3)


d$maf_tertiles= factor(d$maf_tertiles, levels=c("1", "2", "3"), labels=c(paste('MAF<', m1), paste(m1,'< MAF >', m2), paste('MAF>', m2)))

df= arrange(d, pvalue) %>% group_by(maf_tertiles) %>% mutate(exp1= -log10(1:length(pvalue)/length(pvalue)))

ggplot(filter(df, pvalue<0.05), aes(exp1, -log10(pvalue), color= maf_tertiles)) +
  geom_point(size= 0.4) +
scale_color_manual(values= colorBlindBlack8[2:4])+ 
  geom_abline(intercept = 0, slope = 1, alpha = .5) +
labs(colour="") +
theme_cowplot(font_size= 12) +
xlab('Expected (-log10(p-value))') +
ylab('Observed (-log10(p-value))') +
theme(legend.position= 'bottom') +
guides(colour = guide_legend(override.aes = list(size=3)))

```


\pagebreak

# Top genetic variants from each locus  

```{r echo=F}
don= filter(don, pvalue< 1*10**-4)
d1= filter(don, pvalue<5*10**-8)
d1= d1[order(d1$pvalue, decreasing= F), ]
d1$pvalue= format(d1$pvalue, digits= 3)
don$pvalue= format(don$pvalue, digits= 3)

kable(filter(d1, GENE!= '') %>% select(ID, RSID, GENE, TOTALSAMPLESIZE, EAF, BETA, SE, pvalue, Consequence), col.names= c('CHR:POS:REF:EFF', 'RSID', 'Gene', 'N', 'EAF', 'Beta', 'SE', 'P-value', 'Consequence'), digits= 3) #%>% kable_styling(latex_options = c("striped", "scale_down"))


```


Most variants are either associated with birth weight (maternal or fetal effect) and with height or bone mineral density^[[Open Targets](https://genetics.opentargets.org/)].





# Effect vs MAF  



```{r, echo=F, fig.cap= "As expected, beta increases with decreasing minor allele frequency.", fig.align = "center", fig.width= 8, fig.height= 7, dpi= 600}

if (nrow(dg)>1){

(ggplot(dg, aes(MAF, abs(BETA), size= abs(BETA)),  alpha= 0.7) +
geom_point(colour= colorBlindBlack8[4]) +
theme_cowplot(font_size= 12) +
scale_size_continuous(name= 'Absolute Beta', guide= F) +
geom_text_repel(data= dg, aes(label= GENE), guide= F) +
xlab('Minor allele frequency') +
ylab('Absolute effect size')) %>% print()

} else{ 

print('Only one locus identified, check the table.')
plot_comment=''
}

```


# Heterogeneity vs Association P-value  

```{r, echo=F, fig.align = "center", fig.width= 8, fig.height= 7, dpi= 600, fig.caption= 'No pattern between effect size and heterogeneity. Attention should be paid to top hits (the lowest pvalues -> higher heterogeneity).\n Effect sizes aligned to the alphabetically higher allele, so effect direction is coloured simply for aesthetics.'}

z= fread(snakemake@input[[5]], header= T, sep= '\t', select= c('MarkerName', 'Effect', 'P-value', 'HetPVal'))

names(z)= c('ID', 'beta', 'pvalue', 'het_pvalue')
z$ID= gsub(':SNP', '', z$ID)
z$ID= gsub(':INDEL', '', z$ID)

z= inner_join(z, dg, by= 'ID')


if (nrow(z)>1) {

plot_comment= 'No pattern between effect size and heterogeneity. Attention should be paid to the top hit.'

z$Direction= ifelse(z$beta> 0, 'Positive', 'Negative')
ggplot(z, aes(-log10(het_pvalue), -log10(pvalue), size= abs(beta)), alpha= 0.7) +
geom_point(colour= colorBlindBlack8[4]) +
theme_cowplot(font_size= 12) +
scale_size_continuous(name= 'Absolute effect size') +
geom_text_repel(data= z, aes(label= GENE), hjust =1, show.legend = FALSE) +
xlab('-log10(Het pvalue)') +
ylab('-log10(Association pvalue)') +
theme(legend.position="bottom")

} else{
print(paste('Pvalue for heterogeneity: ', z$het_pvalue))
#plot_comment= ''
}

```


# Sugestive genetic variants with HIGH or MODERATE impact  

Genetic variants within 1.5Mb from the top variants, a p-value< 1x10^-4^ and with HIGH or MODERATE are included in this table. Genetic variants were annotated using [Variant Effect Predictor](https://grch37.ensembl.org/Homo_sapiens/Tools/VEP) command line tool.


```{r echo= F}

kable(filter(don, (IMPACT== 'HIGH') | (IMPACT== 'MODERATE')) %>% select(ID, RSID, SYMBOL, TOTALSAMPLESIZE, EAF, BETA, SE, pvalue, Consequence), col.names= c('CHR:POS:REF:EFF', 'RSID', 'Gene', 'N', 'EAF', 'Beta', 'SE', 'P-value', 'Consequence'), digits= 3)

x= readLines(snakemake@input[[2]])
x= x[match('Heritability of phenotype 1', x) + 2]

```

\pagebreak


# Genetic correlations and heritability  

Genetic correlations and heritability using LDScore regression with LD scores obtained from 1KGenomes using a core sample of 1M SNPs from HapMap with high imputation accuracy. `r paste0(x, '.')`


```{marginfigure}
Ideally, calculate LDscores from our sample (MOBAGENETICS) or from a bigger cohort (UKBIOBANK).
```

```{r echo=F, fig.align = "center", fig.width= 5, fig.height= 5, dpi= 600, fig.cap= 'Genetic correlations using LDscore regression.'}

d= fread(snakemake@input[[3]])

d$pheno1= gsub('.txt.sumstats.gz', '', sapply(strsplit(as.character(d$p1), '/'), tail, 1))
d$pheno2= gsub('.txt.sumstats.gz', '', sapply(strsplit(as.character(d$p2), '/'), tail, 1))

d$rg= ifelse(d$rg> 1, 1, ifelse(d$rg< ( -1), -1, d$rg))

maxy= with(d, ifelse(max(rg + 1.96*se)> 1, max(rg + 1.96*se), 1))
miny=with(d, ifelse(min(rg - 1.96*se)< -1, min(rg - 1.96*se), -1))

ggplot(d, aes(pheno2, rg, colour= pheno2)) +
  geom_point(alpha= 0.5) +
geom_errorbar(aes(ymin= I(rg - 1.96*se) , ymax= (rg + 1.96 * se)), width=.2, position=position_dodge(.9)) +
theme_cowplot(font_size= 9) +
scale_fill_manual(values= colorBlindBlack8[2:4]) +
scale_colour_manual(guide= F, values= colorBlindBlack8[2:4]) +
xlab('Phenotype') +
ylab(paste0('R coefficient [95% CI] \n', pheno)) +
ylim(c(miny, maxy)) +
theme(legend.position= 'none')

link= 'https://drive.google.com/drive/folders/101ErlqwE4_iFwZFCTM0QZUtUVwOoOE1L?usp=sharing'

```

# Locus plots

LocusZoom plots for each locus can be found [here](`r link`).

# Forest plots  

Meta-analysis forest plots can be found [here](https://drive.google.com/drive/folders/1q1lk4n2ldB5Nixqv4o3LwX3cA6APJ9nM?usp=sharing).
